### PR TITLE
Feature json checklist 376

### PR DESCRIPTION
**Pull request by rejahrehim**
This PR is part of issue #376

- [ ] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

-  Initial commit for JSON checklist generator 
- Github action to send PR with updated JSON checklist

Current format

```
{
    "categories": {
        "Information Gathering": {
            "id":"WSTG-INFO",
            "tests":[
                {
                "name":"Conduct Search Engine Discovery Reconnaissance for Information Leakage",
                "id":"WSTG-INFO-01",
                "Reference":"https://owasp.org/www-project-web-security-testing-guide/stable/4-Web_Application_Security_Testing/01-Information_Gathering/01-Conduct_Search_Engine_Discovery_Reconnaissance_for_Information_Leakage.html"
                }]
       }
}
```

Need to update `"Objectives": "<Grab objectives>",` and `"CRE_ID": "<CRE_IDn>"`
We have to decide how to handle `Objectives` and `CRE_ID`

### CODE REVIEW

File: .github/workflows/build-json.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+name: Deploy Latest WSTG Json Checklist
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:1]:
```suggestion
name: Deploy Latest JSON Checklist
```

File: .github/workflows/build-json.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  create_pr:
+    name: Create Pull Request
+    runs-on: ubuntu-latest
+    steps:
+    - name: Setup Action
+      uses: actions/checkout@v2
+    - name: Setup Node
+      uses: actions/setup-node@v1
+      with:
+        node-version: 12.x
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:20]: Any reason not to do `13.x` or even `14.x`?

File: .github/workflows/build-json.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    - name: Install dependencies
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:21]:
```suggestion
    - name: Install Dependencies
```

File: .github/workflows/build-json.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      run : |
+        # Setup some env vars etc for future use
+        echo "SRC_BASE="OWASP/wstg@"$(git log -1 --format=format:%h)" >> $GITHUB_ENV
+        echo "BRANCH_STAMP=$(date +"%Y%m%d%H%M%S")" >> $GITHUB_ENV
+        echo "SHORT_DATE=$(date +"%Y-%m-%d")" >> $GITHUB_ENV
+    - name: Create Feature Branch
+      run: |
+        # Checkout what will be the new feature branch
+        git checkout -b $BRANCH_STAMP
+    - name: Run Json Build
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:43]:
```suggestion
    - name: Run JSON Build
```

File: .github/workflows/build-json.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      run: npm run-script jsongen
+
+    - name: Push Feature Branch and Raise PR
+      run: |
+        git remote set-url origin https://$GITHUB_USER:${{ secrets.wstg_deploy_token }}@github.com/OWASP/wstg.git
+        git add checklist/Testing-checklist.json
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:49]:
```suggestion
        git add checklist/checklist.json
```

File: .github/workflows/build-json.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        git commit -m "Publish Latest Json checklist $SHORT_DATE" -m "Updates based on $SRC_BASE"
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:50]:
```suggestion
        git commit -m "Publish Latest JSON checklist $SHORT_DATE" -m "Updates based on $SRC_BASE"
```

File: json/scripts/make-json.sh [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        fi
+    done <<< "$d"
+done
+# Close categories
+echo "    }"  >> testing-checklist.json
+# End Json string
+echo "}"  >> testing-checklist.json
+
+# Move generated file to checklist folder
+mv testing-checklist.json ../../checklist/Testing-checklist.json
```
Comment (kingthorin) [inline json/scripts/make-json.sh:66]:
```suggestion
mv testing-checklist.json ../../checklist/.
```

**Review by kingthorin — CHANGES_REQUESTED** [review]

File: .github/workflows/build-json.yml (hunk 2 above)
Comment (rejahrehim) [reply .github/workflows/build-json.yml]: Updated to 14.x

File: .github/workflows/build-json.yml (hunk 5 above)
Comment (rejahrehim) [reply .github/workflows/build-json.yml]: File name updated to `checklist.json`

File: json/scripts/make-json.sh (hunk 7 above)
Comment (rejahrehim) [reply json/scripts/make-json.sh]: File name updated to `checklist.json`

**Review by kingthorin — APPROVED** [review]
Looks good, thanks ⭐

**Comment by kingthorin** [comment]
Should add a check that the file is actually changed before trying to raise the PR. I've done so elsewhere, will dig up the necessary details in a few hours.

File: .github/workflows/build-json.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    - name: Run JSON Build
+      run: npm run-script jsongen
+
+    - name: Push Feature Branch and Raise PR
+      run: |
+        git remote set-url origin https://$GITHUB_USER:${{ secrets.wstg_deploy_token }}@github.com/OWASP/wstg.git
+        git add checklist/checklist.json
+        git commit -m "Publish Latest JSON checklist $SHORT_DATE" -m "Updates based on $SRC_BASE"
+        git push --set-upstream origin $BRANCH_STAMP
+        hub pull-request --no-edit
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:52]:
```suggestion
        ## If there are changes: comment, commit, PR
        if ! git diff-index --quiet HEAD --; then 
          git remote set-url origin https://$GITHUB_USER:${{ secrets.wstg_deploy_token }}@github.com/OWASP/wstg.git
          git add checklist/checklist.json
          git commit -m "Publish Latest JSON checklist $SHORT_DATE" -m "Updates based on $SRC_BASE"
          git push --set-upstream origin $BRANCH_STAMP
          hub pull-request --no-edit
        fi
```

File: .github/workflows/build-json.yml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        echo "SRC_BASE="OWASP/wstg@"$(git log -1 --format=format:%h)" >> $GITHUB_ENV
+        echo "BRANCH_STAMP=$(date +"%Y%m%d%H%M%S")" >> $GITHUB_ENV
+        echo "SHORT_DATE=$(date +"%Y-%m-%d")" >> $GITHUB_ENV
+    - name: Create Feature Branch
+      run: |
+        # Checkout what will be the new feature branch
+        git checkout -b $BRANCH_STAMP
+    - name: Run JSON Build
+      run: npm run-script jsongen
+
```
Comment (kingthorin) [inline .github/workflows/build-json.yml:45]:
```suggestion
```

**Review by kingthorin — COMMENTED** [review]
Looks good, thanks ⭐

**Comment by ThunderSon** [comment]
Any artifact that can be reviewed from my side?

File: json/scripts/make-json.sh [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        if [ ${FOLD[0]} -gt 0 ]
+        then
+            category=${FOLD[1]}
+            # Start category sub section
+            # Add coma from the second entry of the list
+            if [ ${FOLD[0]} -gt 1 ];then
+                echo "        ,\"${category%?}\": {" | tr '_' ' '  >> checklist.json
+            else
+                echo "        \"${category%?}\": {" | tr '_' ' '  >> checklist.json
+            fi
```
Comment (ThunderSon) [inline json/scripts/make-json.sh:24]: Wouldn't setting this condition at the closing end better in terms of formatting?

File: json/scripts/make-json.sh [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            count=0
+            for file in $d*.md; do
+                # Remove README
+                if [[ $file != *"README.md" ]];then
+                    # Add coma from the second entry of the list
+                    if [ $count -gt 0 ];then
+                        echo "                ,{"  >> checklist.json
+                    else
+                        echo "                {"  >> checklist.json
+                    fi
```
Comment (ThunderSon) [inline json/scripts/make-json.sh:39]: Same comment for this

File: json/scripts/make-json.sh [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+                    # Get test ID, test name and reference link from the file
+                    tid=`cat $file | grep "|WSTG-.*" | cut -d "|" -f 2`
+                    tname=$(head -n 1 $file)
```
Comment (ThunderSon) [inline json/scripts/make-json.sh:42]: From a small research, I believe this is a better format.
```suggestion
                    read -r tname < $file
```

**Review by ThunderSon — COMMENTED** [review]
Basic comments.
Another question, can this JSON file be somewhat done differently than just parsing line by line like this? Like a template that gets injected with the required data. Not a hard requirement now, but I can see this becoming bad for however wanna push this work forward later down the line.

File: json/scripts/make-json.sh (hunk 10 above)
Comment (rejahrehim) [reply json/scripts/make-json.sh:24]: Tried, but getting the number of entries is bit complected.

File: json/scripts/make-json.sh (hunk 11 above)
Comment (rejahrehim) [reply json/scripts/make-json.sh:39]: Tried, but getting the number of entries is bit complected.  Also, We have to exclude `README.md` Files.

**Review by kingthorin — APPROVED** [review]

**Review by ThunderSon — APPROVED** [review]
Thanks! :)

**Pull request merged**